File: 0048-rotate-image/0048-rotate-image.java
class Solution {
    public void rotate(int[][] matrix) {
        int n = matrix.length;

        // process layer by layer
        for (int layer = 0; layer < n / 2; layer++) {
            int first = layer;
            int last = n - 1 - layer;

            for (int i = first; i < last; i++) {
                int top = matrix[first][i];

                int right = matrix[i][last];

                int bottom = matrix[last][n - 1 - i];

                int left = matrix[n - 1 - i][first];

                // left -> top
                matrix[first][i] = matrix[n - 1 - i][first];

                // bottom -> left
                matrix[n - 1 - i][first] = matrix[last][n - 1 - i];

                // right -> bottom
                matrix[last][n - 1 - i] = matrix[i][last];

                // top -> right
                matrix[i][last] = top;
            }

        }
    }
}

File: 0048-rotate-image/0048-rotate-image.py
class Solution:
    def rotate(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        n = len(matrix)
        for i in range(n):
            for j in range(i+1, n):
                matrix[j][i], matrix[i][j] = matrix[i][j], matrix[j][i]
        
        for i in range(n):
            for j in range(n // 2):
                print(i, j)
                matrix[i][j], matrix[i][-j - 1] = matrix[i][-j - 1], matrix[i][j]
        return matrix